### PR TITLE
fix: translate command attributes

### PR DESCRIPTION
**Pull request by williamboman**

### CODE REVIEW

**Comment by williamboman** [comment]
Hey @ranjithshegde, I found some issues while running the `0.7` branch. The `rust-tools.nvim` plugin provides command attributes [like so](https://github.com/simrat39/rust-tools.nvim/blob/061977c0ddebd561ae3dd89e3900aeccf96c1946/lua/rust-tools/lsp.lua#L19) in some cases. Regardless of if that's according to the intended API, it works in `master`. I figured I'd open a PR directly as I feel slightly responsible for pointing you in the direction of translating commands :D. Hopefully after this the 0.7 branch is good to go.

**Comment by williamboman** [comment]
One sec, let me give this a little more testing

File: lua/lspconfig/util.lua [hunk 1]
```diff
@@ -87,8 +87,8 @@ end
 function M.create_module_commands(module_name, commands)
   for command_name, def in pairs(commands) do
     local opts = M._parse_user_command_options(def)
-    api.nvim_create_user_command(command_name, function(...)
-      require('lspconfig')[module_name].commands[command_name][1](...)
+    api.nvim_create_user_command(command_name, function(info)
+      require('lspconfig')[module_name].commands[command_name][1](unpack(info.fargs))
```
Comment (williamboman) [inline lua/lspconfig/util.lua:91]: The old way would only provide `fargs` like so: `require'lspconfig'[%q].commands[%q][1](<f-args>)`. The Lua equivalent of this is simply `unpack(info.fargs)`!

**Comment by ranjithshegde** [comment]
Thanks for this. After you pointed it out in the review, this was my concern, that I dont know the comprehensive list of arguments that `:commands` takes as I am only familiar with some of them!

File: lua/lspconfig/util.lua [hunk 2]
```diff
@@ -63,22 +63,32 @@ local opts_aliases = {
...
+---@param command_definition table<string | integer, any>
+function M._parse_user_command_options(command_definition)
+  ---@type table<string, string | boolean | number>
+  local opts = {}
+  for k, v in pairs(command_definition) do
+    if type(k) == 'string' then
+      local attribute = k.gsub(k, '^%-+', '')
+      opts[opts_aliases[attribute] or attribute] = v
+    elseif type(k) == 'number' and type(v) == 'string' and v:match '^%-' then
+      -- Splits strings like "-narsg=* -complete=customlist,v:lua.something" into { "-nargs=*", "-complete=customlist,v:lua.something" }
```
Comment (williamboman) [inline lua/lspconfig/util.lua]:
```suggestion
      -- Splits strings like "-nargs=* -complete=customlist,v:lua.something" into { "-nargs=*", "-complete=customlist,v:lua.something" }
```

**Comment by williamboman** [comment]
This fixes all immediate issues I had when running `0.7`. I feel like if anything's wrong it'd explode early, so I feel pretty comfortable with this!

**Comment by williamboman** [comment]
Ok that was the last commit, I promise 🤞

**Comment by ranjithshegde** [comment]
Looks great!
I suppose if there are edge cases, we can fix it as we go. At the moment `-f/n/args` and `string` options are handled fully!

Shall I merge this?

**Comment by williamboman** [comment]
> Looks great! I suppose if there are edge cases, we can fix it as we go. At the moment `-f/n/args` and `string` options are 
handled fully!

Yeah if something pops up later we can just add a regression test for it and deal with it then!

> Shall I merge this?

Let's do it! Hopefully the 0.7 branch can get merged upstream soon as well, great work :+1:

**Comment by ranjithshegde** [comment]
@williamboman 
For some reason this got fast-forwarded the wrong way is not in my local or remote. Really sorry about this. No clue what happened.
I can copy paste the diffs, or you can resubmit the PR and I will merge again.

What do you say?

**Comment by williamboman** [comment]
I deleted the fork already. Reflog to the rescue :D!

```diff
diff --git a/lua/lspconfig/util.lua b/lua/lspconfig/util.lua
index 73b5c88..6bcd0a7 100644
--- a/lua/lspconfig/util.lua
+++ b/lua/lspconfig/util.lua
@@ -63,22 +63,32 @@ local opts_aliases = {
   ['description'] = 'desc',
 }
 
-function M.create_module_commands(module_name, commands)
-  for command_name, def in pairs(commands) do
-    local opts = {}
-    -- Insert attributes.
-    for k, v in pairs(def) do
-      if type(k) == 'string' then
-        opts[opts_aliases[k] or k] = v
-      elseif type(k) == 'number' and type(v) == 'string' and v:match '^%-' then
-        local option, value = unpack(vim.split(v, '=', { plain = true }))
-        option = option.gsub(v, '^%-+', '')
-        opts[opts_aliases[option] or option] = value or true
+---@param command_definition table<string | integer, any>
+function M._parse_user_command_options(command_definition)
+  ---@type table<string, string | boolean | number>
+  local opts = {}
+  for k, v in pairs(command_definition) do
+    if type(k) == 'string' then
+      local attribute = k.gsub(k, '^%-+', '')
+      opts[opts_aliases[attribute] or attribute] = v
+    elseif type(k) == 'number' and type(v) == 'string' and v:match '^%-' then
+      -- Splits strings like "-nargs=* -complete=customlist,v:lua.something" into { "-nargs=*", "-complete=customlist,v:lua.something" }
+      for _, command_attribute in ipairs(vim.split(v, '%s')) do
+        -- Splits attribute into a key-value pair, like "-nargs=*" to { "-nargs", "*" }
+        local attribute, value = unpack(vim.split(command_attribute, '=', { plain = true }))
+        attribute = attribute.gsub(attribute, '^%-+', '')
+        opts[opts_aliases[attribute] or attribute] = value or true
       end
     end
-    -- The command definition.
-    api.nvim_create_user_command(command_name, function(...)
-      require('lspconfig')[module_name].commands[command_name][1](...)
+  end
+  return opts
+end
+
+function M.create_module_commands(module_name, commands)
+  for command_name, def in pairs(commands) do
+    local opts = M._parse_user_command_options(def)
+    api.nvim_create_user_command(command_name, function(info)
+      require('lspconfig')[module_name].commands[command_name][1](unpack(info.fargs))
     end, opts)
   end
 end
diff --git a/plugin/lspconfig.lua b/plugin/lspconfig.lua
index fe32763..774e40b 100644
--- a/plugin/lspconfig.lua
+++ b/plugin/lspconfig.lua
@@ -9,8 +9,11 @@ if vim.fn.has 'nvim-0.7' ~= 1 then
   return
 end
 
+local util = require 'lspconfig.util'
+local configs = require 'lspconfig.configs'
+
 _G.lsp_complete_configured_servers = function()
-  return table.concat(vim.tbl_keys(require 'lspconfig.configs'), '\n')
+  return table.concat(vim.tbl_keys(configs), '\n')
 end
 
 _G.lsp_get_active_client_ids = function()
@@ -28,7 +31,7 @@ local get_clients_from_cmd_args = function(arg)
     result[id] = vim.lsp.get_client_by_id(tonumber(id))
   end
   if vim.tbl_isempty(result) then
-    return require('lspconfig.util').get_managed_clients()
+    return util.get_managed_clients()
   end
   return vim.tbl_values(result)
 end
@@ -44,8 +47,8 @@ end, {
 vim.api.nvim_create_user_command('LspStart', function(info)
   local server_name = info.args[1] ~= '' and info.args
   if server_name then
-    if require('lspconfig.configs')[server_name] then
-      require('lspconfig.configs')[server_name].launch()
+    if configs[server_name] then
+      configs[server_name].launch()
     end
   else
     local other_matching_configs = require('lspconfig.util').get_other_matching_providers(vim.bo.filetype)
@@ -62,7 +65,7 @@ vim.api.nvim_create_user_command('LspRestart', function(info)
   for _, client in ipairs(get_clients_from_cmd_args(info.args)) do
     client.stop()
     vim.defer_fn(function()
-      require('lspconfig.configs')[client.name].launch()
+      configs[client.name].launch()
     end, 500)
   end
 end, {
diff --git a/test/lspconfig_spec.lua b/test/lspconfig_spec.lua
index 163bf88..ff9ceb5 100644
--- a/test/lspconfig_spec.lua
+++ b/test/lspconfig_spec.lua
@@ -190,6 +190,41 @@ describe('lspconfig', function()
         ]])
       end)
     end)
+
+    describe('user commands', function()
+      it('should translate command definition to nvim_create_user_command options', function()
+        eq(
+          {
+            nargs = '*',
+            complete = 'custom,v:lua.some_global',
+          },
+          exec_lua [[
+            local util = require("lspconfig.util")
+            return util._parse_user_command_options({
+              function () end,
+              "-nargs=* -complete=custom,v:lua.some_global"
+            })
+          ]]
+        )
+
+        eq(
+          {
+            desc = 'My awesome description.',
+            nargs = '*',
+            complete = 'custom,v:lua.another_global',
+          },
+          exec_lua [[
+            local util = require("lspconfig.util")
+            return util._parse_user_command_options({
+              function () end,
+              ["-nargs"] = "*",
+              "-complete=custom,v:lua.another_global",
+              description = "My awesome description."
+            })
+          ]]
+        )
+      end)
+    end)
   end)
   describe('config', function()
     it('normalizes user, server, and base default configs', function()
```

**Comment by williamboman** [comment]
<details><summary>So this was the original diff right before merge:</summary>

```diff
diff --git a/lua/lspconfig/util.lua b/lua/lspconfig/util.lua
index 73b5c88..6bcd0a7 100644
--- a/lua/lspconfig/util.lua
+++ b/lua/lspconfig/util.lua
@@ -63,22 +63,32 @@ local opts_aliases = {
   ['description'] = 'desc',
 }
 
-function M.create_module_commands(module_name, commands)
-  for command_name, def in pairs(commands) do
-    local opts = {}
-    -- Insert attributes.
-    for k, v in pairs(def) do
-      if type(k) == 'string' then
-        opts[opts_aliases[k] or k] = v
-      elseif type(k) == 'number' and type(v) == 'string' and v:match '^%-' then
-        local option, value = unpack(vim.split(v, '=', { plain = true }))
-        option = option.gsub(v, '^%-+', '')
-        opts[opts_aliases[option] or option] = value or true
+---@param command_definition table<string | integer, any>
+function M._parse_user_command_options(command_definition)
+  ---@type table<string, string | boolean | number>
+  local opts = {}
+  for k, v in pairs(command_definition) do
+    if type(k) == 'string' then
+      local attribute = k.gsub(k, '^%-+', '')
+      opts[opts_aliases[attribute] or attribute] = v
+    elseif type(k) == 'number' and type(v) == 'string' and v:match '^%-' then
+      -- Splits strings like "-nargs=* -complete=customlist,v:lua.something" into { "-nargs=*", "-complete=customlist,v:lua.something" }
+      for _, command_attribute in ipairs(vim.split(v, '%s')) do
+        -- Splits attribute into a key-value pair, like "-nargs=*" to { "-nargs", "*" }
+        local attribute, value = unpack(vim.split(command_attribute, '=', { plain = true }))
+        attribute = attribute.gsub(attribute, '^%-+', '')
+        opts[opts_aliases[attribute] or attribute] = value or true
       end
     end
-    -- The command definition.
-    api.nvim_create_user_command(command_name, function(...)
-      require('lspconfig')[module_name].commands[command_name][1](...)
+  end
+  return opts
+end
+
+function M.create_module_commands(module_name, commands)
+  for command_name, def in pairs(commands) do
+    local opts = M._parse_user_command_options(def)
+    api.nvim_create_user_command(command_name, function(info)
+      require('lspconfig')[module_name].commands[command_name][1](unpack(info.fargs))
     end, opts)
   end
 end
diff --git a/plugin/lspconfig.lua b/plugin/lspconfig.lua
index fe32763..774e40b 100644
--- a/plugin/lspconfig.lua
+++ b/plugin/lspconfig.lua
@@ -9,8 +9,11 @@ if vim.fn.has 'nvim-0.7' ~= 1 then
   return
 end
 
+local util = require 'lspconfig.util'
+local configs = require 'lspconfig.configs'
+
 _G.lsp_complete_configured_servers = function()
-  return table.concat(vim.tbl_keys(require 'lspconfig.configs'), '\n')
+  return table.concat(vim.tbl_keys(configs), '\n')
 end
 
 _G.lsp_get_active_client_ids = function()
@@ -28,7 +31,7 @@ local get_clients_from_cmd_args = function(arg)
     result[id] = vim.lsp.get_client_by_id(tonumber(id))
   end
   if vim.tbl_isempty(result) then
-    return require('lspconfig.util').get_managed_clients()
+    return util.get_managed_clients()
   end
   return vim.tbl_values(result)
 end
@@ -44,8 +47,8 @@ end, {
 vim.api.nvim_create_user_command('LspStart', function(info)
   local server_name = info.args[1] ~= '' and info.args
   if server_name then
-    if require('lspconfig.configs')[server_name] then
-      require('lspconfig.configs')[server_name].launch()
+    if configs[server_name] then
+      configs[server_name].launch()
     end
   else
     local other_matching_configs = require('lspconfig.util').get_other_matching_providers(vim.bo.filetype)
@@ -62,7 +65,7 @@ vim.api.nvim_create_user_command('LspRestart', function(info)
   for _, client in ipairs(get_clients_from_cmd_args(info.args)) do
     client.stop()
     vim.defer_fn(function()
-      require('lspconfig.configs')[client.name].launch()
+      configs[client.name].launch()
     end, 500)
   end
 end, {
diff --git a/test/lspconfig_spec.lua b/test/lspconfig_spec.lua
index 163bf88..ff9ceb5 100644
--- a/test/lspconfig_spec.lua
+++ b/test/lspconfig_spec.lua
@@ -190,6 +190,41 @@ describe('lspconfig', function()
         ]])
       end)
     end)
+
+    describe('user commands', function()
+      it('should translate command definition to nvim_create_user_command options', function()
+        eq(
+          {
+            nargs = '*',
+            complete = 'custom,v:lua.some_global',
+          },
+          exec_lua [[
+            local util = require("lspconfig.util")
+            return util._parse_user_command_options({
+              function () end,
+              "-nargs=* -complete=custom,v:lua.some_global"
+            })
+          ]]
+        )
+
+        eq(
+          {
+            desc = 'My awesome description.',
+            nargs = '*',
+            complete = 'custom,v:lua.another_global',
+          },
+          exec_lua [[
+            local util = require("lspconfig.util")
+            return util._parse_user_command_options({
+              function () end,
+              ["-nargs"] = "*",
+              "-complete=custom,v:lua.another_global",
+              description = "My awesome description."
+            })
+          ]]
+        )
+      end)
+    end)
   end)
   describe('config', function()
     it('normalizes user, server, and base default configs', function()
```

</details>

<details><summary>This is the diff now (slightly changed, but not the parts I changed)</summary>

```diff
diff --git a/lua/lspconfig/util.lua b/lua/lspconfig/util.lua
index 73b5c88..6bcd0a7 100644
--- a/lua/lspconfig/util.lua
+++ b/lua/lspconfig/util.lua
@@ -63,22 +63,32 @@ local opts_aliases = {
   ['description'] = 'desc',
 }
 
-function M.create_module_commands(module_name, commands)
-  for command_name, def in pairs(commands) do
-    local opts = {}
-    -- Insert attributes.
-    for k, v in pairs(def) do
-      if type(k) == 'string' then
-        opts[opts_aliases[k] or k] = v
-      elseif type(k) == 'number' and type(v) == 'string' and v:match '^%-' then
-        local option, value = unpack(vim.split(v, '=', { plain = true }))
-        option = option.gsub(v, '^%-+', '')
-        opts[opts_aliases[option] or option] = value or true
+---@param command_definition table<string | integer, any>
+function M._parse_user_command_options(command_definition)
+  ---@type table<string, string | boolean | number>
+  local opts = {}
+  for k, v in pairs(command_definition) do
+    if type(k) == 'string' then
+      local attribute = k.gsub(k, '^%-+', '')
+      opts[opts_aliases[attribute] or attribute] = v
+    elseif type(k) == 'number' and type(v) == 'string' and v:match '^%-' then
+      -- Splits strings like "-nargs=* -complete=customlist,v:lua.something" into { "-nargs=*", "-complete=customlist,v:lua.something" }
+      for _, command_attribute in ipairs(vim.split(v, '%s')) do
+        -- Splits attribute into a key-value pair, like "-nargs=*" to { "-nargs", "*" }
+        local attribute, value = unpack(vim.split(command_attribute, '=', { plain = true }))
+        attribute = attribute.gsub(attribute, '^%-+', '')
+        opts[opts_aliases[attribute] or attribute] = value or true
       end
     end
-    -- The command definition.
-    api.nvim_create_user_command(command_name, function(...)
-      require('lspconfig')[module_name].commands[command_name][1](...)
+  end
+  return opts
+end
+
+function M.create_module_commands(module_name, commands)
+  for command_name, def in pairs(commands) do
+    local opts = M._parse_user_command_options(def)
+    api.nvim_create_user_command(command_name, function(info)
+      require('lspconfig')[module_name].commands[command_name][1](unpack(info.fargs))
     end, opts)
   end
 end
diff --git a/test/lspconfig_spec.lua b/test/lspconfig_spec.lua
index 163bf88..ff9ceb5 100644
--- a/test/lspconfig_spec.lua
+++ b/test/lspconfig_spec.lua
@@ -190,6 +190,41 @@ describe('lspconfig', function()
         ]])
       end)
     end)
+
+    describe('user commands', function()
+      it('should translate command definition to nvim_create_user_command options', function()
+        eq(
+          {
+            nargs = '*',
+            complete = 'custom,v:lua.some_global',
+          },
+          exec_lua [[
+            local util = require("lspconfig.util")
+            return util._parse_user_command_options({
+              function () end,
+              "-nargs=* -complete=custom,v:lua.some_global"
+            })
+          ]]
+        )
+
+        eq(
+          {
+            desc = 'My awesome description.',
+            nargs = '*',
+            complete = 'custom,v:lua.another_global',
+          },
+          exec_lua [[
+            local util = require("lspconfig.util")
+            return util._parse_user_command_options({
+              function () end,
+              ["-nargs"] = "*",
+              "-complete=custom,v:lua.another_global",
+              description = "My awesome description."
+            })
+          ]]
+        )
+      end)
+    end)
   end)
   describe('config', function()
     it('normalizes user, server, and base default configs', function()
```

</details>

**Comment by ranjithshegde** [comment]
Thanks for this. Really want to how how the fast-forward got borked!
Its applied and committed now
